form/create_custom_choice_type.rst

```diff
@@ -0,0 +1,273 @@
+    a third argument ``$vary`` can be used as array to pass any value that
+    makes the loaded choices different.
+
+Creating a Type With Dynamic Choices
+------------------------------------
+
+When loading choices is complex, a callback is not enough and a "real" service
+is needed. Fortunately, the Form component provides a
+:class:`Symfony\\Component\\Form\\ChoiceList\\Loader\\ChoiceLoaderInterface`.
+You can pass any instance to the ``choice_loader`` option to handle things
+any way you need. For example, you could leverage this new power to load
+categories from an HTTP API. The easiest way is to extend the
+:class:`Symfony\\Component\\Form\\ChoiceList\\Loader\\AbstractChoiceLoader`
+class, which already implements the interface and avoids triggering your logic
+when it is not needed (i.e when the form is submitted empty and valid).
+This could look like this::
+
+    // src/Form/ChoiceList/AcmeCategoryLoader.php.
+    namespace App\Form\ChoiceList;
+
+    use App\Api\AcmeApi;
+    use Symfony\Component\Form\ChoiceList\Loader\AbstractChoiceLoader;
+
+    class AcmeCategoryLoader extends AbstractChoiceLoader;
+    {
+        // This must be passed by the type
+        // this loader won't be registered as service
+        private $api;
+        // define more options if needed
+        private $someOption;
+
+        public function __construct(AcmeApi $api, string $someOption)
+        {
+            $this->api = $api;
+            $this->someOption = $someOption;
+        }
+
+        protected function loadChoices(): iterable
+        {
+            return $this->api->loadCategories($this->someOption));
+        }
+
+        protected function doLoadChoicesForValues(array $values): array
+        {
+            return $this->api->loadCategoriesForNames($values, $this->someOption);
+        }
+
+        protected function doLoadValuesForChoices(array $choices): array
+        {
+            $values = [];
+
+            // compute string values that must be submitted
+
+            return $values;
+        }
+    }
+
+Here we implement three protected methods:
+
+`loadChoices(): iterable`
+
+    This method is abstract and is the only one that needs to be implemented,
+    it is called when the list is fully loaded (i.e when rendering the view).
+    It must return an array or a traversable object, keys are default labels
+    unless the :ref:`choice_label <reference-form-choice-label>` option is
+    defined.
+    Choices can be grouped with keys as group name and nested iterable choices
+    in alternative to the :ref:`group_by <reference-form-group-by>` option.
+
+`doLoadChoicesForValues(array $values): array`
+
+    Optional, to improve performance this method is called when the data is
+    submitted. You can then load the choices partially, by using the submitted
+    values passed as only argument.
+    The list is fully loaded by default.
+
+`doLoadValuesForChoices(array $choices): array`
+
+    Optional, in alternative to the
```

Comment:
```suggestion
    Optional, as alternative to the
```

(not very sure about this one)

Reply:
I've fixed, maybe @weaverryan missed it or I will revert. You made me doubt :D.